File: framework/languages/dotnet/src/Zlink.Framework/Contracts/Streams/IZLinkSessionPacketHandler.cs
namespace Zlink.Framework.Contracts.Streams;

public interface IZLinkSessionHandlerRegistry
{
    void AddHandler<THandler>()
        where THandler : class;

    void AddHandler<THandler>(string packetName)
        where THandler : class;

    ValueTask<bool> TryHandleAsync(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload,
        CancellationToken cancellationToken = default);
}

public interface IZLinkSessionPacketHandler<in TSessionContext, TMessage>
{
    ValueTask HandleAsync(
        TSessionContext context,
        ZLinkSessionDispatchContext dispatch,
        TMessage message,
        CancellationToken cancellationToken);
}

File: framework/languages/node/packages/framework/src/contracts/Streams/IZLinkSessionPacketHandler.ts
import type { Type, ZLinkMessage } from '../Common';
import type { ZLinkSessionDispatchContext } from './IZLinkSession';

export interface ZLinkSessionPacketHandler<TSessionContext, TMessage = ZLinkMessage> {
  handle(context: TSessionContext, dispatch: ZLinkSessionDispatchContext, message: TMessage): Promise<void>;
}

export interface ZLinkSessionHandlerRegistry {
  addHandler<THandler>(handlerType: Type<THandler>): this;
  tryHandle(dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage): Promise<boolean>;
}
